zkevm-circuits/src/state_circuit/state.rs

```diff
@@ -98,453 +75,327 @@ pub struct Config<
+
+            let s_enable = meta.query_fixed(s_enable, Rotation::cur());
+            let is_write = meta.query_advice(is_write, Rotation::cur());
+            let q_read = one.clone() - is_write;
+            let key2 = meta.query_advice(keys[2], Rotation::cur());
+            let key4 = meta.query_advice(keys[4], Rotation::cur());
+
+            // 0. Unused keys are 0
+            cb.require_zero("key2 is 0", key2);
+            cb.require_zero("key4 is 0", key4);
+
+            // 1. First access for a set of all keys
+            //
+            // The first stack operation in a stack position is always a write (can't
+            // read if it isn't written before)
+            //
+            // When the set of all keys changes (first access of a stack position in a call)
+            // - It must be a WRITE
+            cb.require_zero(
+                "if address changes, operation is always a write",
+                q_not_all_keys_same(meta) * q_read,
+            );
+            cb.gate(s_enable * q_stack(meta))
         });
 
-        meta.create_gate("Storage operation", |meta| {
-            let q_storage_not_first = q_storage_not_first(meta);
-            let address_diff = {
-                let address_prev = meta.query_advice(address, Rotation::prev());
-                let address_cur = meta.query_advice(address, Rotation::cur());
-                address_cur - address_prev
-            };
-
-            let storage_key_diff = {
-                let storage_key_prev =
-                    meta.query_advice(storage_key, Rotation::prev());
-                let storage_key_cur =
-                    meta.query_advice(storage_key, Rotation::cur());
-                storage_key_cur - storage_key_prev
-            };
-
-            let value_cur = meta.query_advice(value, Rotation::cur());
-            let value_prev_cur = meta.query_advice(value_prev, Rotation::cur());
-            let value_prev_prev =
-                meta.query_advice(value_prev, Rotation::prev());
-            let flag = meta.query_advice(flag, Rotation::cur());
-
-            // flag == 0 or 1
-            // (flag) * (1 - flag)
-            let bool_check_flag = flag.clone() * (one.clone() - flag.clone());
-
-            // If flag == 0 (read), and rw_counter != 0, value_prev == value_cur
-            let value_previous = meta.query_advice(value, Rotation::prev());
-            let q_read = one.clone() - flag.clone();
+        // 2. stack_ptr in range
+        meta.lookup_any("Stack address in allowed range", |meta| {
+            let q_stack = q_stack(meta);
+            let address_cur = meta.query_advice(address, Rotation::cur());
+            let stack_address_table_zero =
+                meta.query_fixed(stack_address_table_zero, Rotation::cur());
 
-            let padding = meta.query_advice(padding, Rotation::cur());
-            let is_not_padding = one.clone() - padding;
+            vec![(q_stack * address_cur, stack_address_table_zero)]
```

Comment:
Same question as before: Usage of s_enable in the lookup?

Reply:
same